core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java

```diff
@@ -598,13 +632,42 @@ private static void closeStatement(PreparedStatement preparedStatement) {
         if (preparedStatement != null) {
             try {
                 preparedStatement.close();
+            } catch (SQLRecoverableException ex) {
+                handleSQLRecoverableException(ex);
             } catch (SQLException e) {
                 log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
             }
         }
 
     }
 
+    private static void handleSQLRecoverableException(SQLRecoverableException ex){
+		Connection dbConnection;
+    	try {
+	    	if(dataSource != null){
+	    		dbConnection = getDBConnection(dataSource);
+				handleSQLRecoverableException(dbConnection, ex);
+	    	}
+    	} catch (SQLException e) {
+            log.error("Error occurred during SQLRecoverableException handling." + e.getMessage());
+    	}
+    }
+    
+    	private static void handleSQLRecoverableException(Connection dbConnection, SQLRecoverableException ex) {
```

Comment:
It seems like handleSQLRecoverableException() method is trying to handle a recoverable situation of a close() method invocation. IMO if the connection is already closed we might not need to handle it again. Can you please explain your intension here?